README

                               GtkGuitune
                         --==#############==--

                   Copyright (C) 1999 Florian Berger
                   
                email: florian.berger@jk.uni-linz.ac.at
                  

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License Version 2 as
    published by the Free Software Foundation;

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.


What is gtkguitune ?
--------------------
   GtkGuitune is a program for tuning guitars (and other instruments).
   It is actually an oscilloscope which counts the passes between two
   trigger levels per time, calculates the frequency and shows the
   tone on a tone-scale.

Resources
---------
   You need to have gtk and gtk-- (1.2.1 or later) installed
   
How to build
------------
   (1) make
   (2) make install

   ...ready.


debian/changelog
gtkguitune (0.8-6) UNRELEASED; urgency=medium

  * Trim trailing whitespace.
  * Bump debhelper from deprecated 7 to 10.
  * Use secure URI in Vcs control header Vcs-Git.
  * Rely on pre-initialized dpkg-architecture variables.
  * Remove constraints unnecessary since buster:
    + Build-Depends: Drop versioned constraint on debhelper.

 -- Debian Janitor <janitor@jelmer.uk>  Fri, 24 Apr 2020 08:04:01 +0000

gtkguitune (0.8-5) unstable; urgency=low

  * set source format to 3.0 (Closes: #638292)

 -- Antonin Kral <A.Kral@sh.cvut.cz>  Thu, 18 Aug 2011 20:47:28 +0200

gtkguitune (0.8-4) unstable; urgency=low

  * switched patch management to gbp-pq and quilt
  * FTBFS on hurd-i386 (Closes: #637549)
  * fixed various lintian warnings

 -- Antonin Kral <A.Kral@sh.cvut.cz>  Wed, 17 Aug 2011 11:37:21 +0200

gtkguitune (0.8-3) unstable; urgency=low

  * use dpatch for patch management, including FTBFS on GNU/kFreeBSD
    refer to bug #336898
  * Included patch by Olivier Robert to support syllables as alternative note
    names (Closes: #556449)
  * cleanup of several lintian warnings

 -- Antonin Kral <A.Kral@sh.cvut.cz>  Wed, 23 Dec 2009 08:49:00 +0100

gtkguitune (0.8-2) unstable; urgency=low

  * add oss-compat, related to bug #514659

 -- Antonin Kral <A.Kral@sh.cvut.cz>  Tue, 10 Feb 2009 11:32:45 +0100

gtkguitune (0.8-1) unstable; urgency=low

  * New upstream release (Closes: #509460)
  * introduced hack into rules to prevent linking against useless libraries
    introduced by gtkmm pkg-config
  * fixed lintian warning debian-rules-ignores-make-clean-error
  * fixed lintian warning copyright-without-copyright-notice
  * fixed lintian warning

 -- Antonin Kral <A.Kral@sh.cvut.cz>  Sun, 08 Feb 2009 14:13:03 +0100

gtkguitune (0.7-7.1) unstable; urgency=medium

  * Non-maintainer upload.
  * Fix FTBFS with GCC 4.3 (Closes: #417220).

 -- Luk Claes <luk@debian.org>  Sun, 09 Mar 2008 15:12:12 +0000

gtkguitune (0.7-7) unstable; urgency=low

  * recompile against libgtkmm1.2-0c2a. Closes: #356282
  * updated postal address of the Free Software Foundation (FSF)

 -- Antonin Kral <A.Kral@sh.cvut.cz>  Sat, 11 Mar 2006 16:27:33 +0100

gtkguitune (0.7-6) unstable; urgency=low

  * acknowleadgement of NMU, many thanks to Matej
  * Closes: #342533
  * Closes: #336898

 -- Antonin Kral <A.Kral@sh.cvut.cz>  Wed,  1 Mar 2006 09:57:47 +0100

gtkguitune (0.7-5.1) unstable; urgency=medium

  * Non-maintainer upload.
  * Recompile against libgtkmm1.2-0c2a.  Closes: #342533.
  * Port to GNU/kFreeBSD; thanks to Aurelien Jarno for the patch.
    Closes: #336898.
  * debian/gtkguitune.1x: Fix groff warnings by using named characters.

 -- Matej Vela <vela@debian.org>  Sat, 28 Jan 2006 13:28:11 +0100

gtkguitune (0.7-5) unstable; urgency=low

  * recompiled against libsigc++0c2, Closes #317930

 -- Antonin Kral <A.Kral@sh.cvut.cz>  Sun, 31 Jul 2005 08:56:07 +0200

gtkguitune (0.7-4) unstable; urgency=low

  * Changed menu to follow "quotation" policy

 -- Antonin Kral <A.Kral@sh.cvut.cz>  Tue, 10 Aug 2004 11:14:07 +0200

gtkguitune (0.7-3) unstable; urgency=low

  * Moved from Apps/Tools to Apps/Sound (Closes #227983)

 -- Antonin Kral <A.Kral@sh.cvut.cz>  Thu, 22 Jan 2004 18:49:09 +0100

gtkguitune (0.7-2) unstable; urgency=low

  * Added more build-deps

 -- Antonin Kral <A.Kral@sh.cvut.cz>  Wed, 17 Sep 2003 08:59:10 +0200

gtkguitune (0.7-1) unstable; urgency=low

  * Initial Release.

 -- Antonin Kral <A.Kral@sh.cvut.cz>  Sat,  6 Sep 2003 17:25:38 +0200


config.h
/* config.h.  Generated from config.h.in by configure.  */
/* config.h.in.  Generated from configure.in by autoheader.  */

/* Define to 1 if you have the <fcntl.h> header file. */
#define HAVE_FCNTL_H 1

/* Define to 1 if you have the <inttypes.h> header file. */
#define HAVE_INTTYPES_H 1

/* Define to 1 if you have the <memory.h> header file. */
#define HAVE_MEMORY_H 1

/* Define to 1 if you have the <stdint.h> header file. */
#define HAVE_STDINT_H 1

/* Define to 1 if you have the <stdlib.h> header file. */
#define HAVE_STDLIB_H 1

/* Define to 1 if you have the <strings.h> header file. */
#define HAVE_STRINGS_H 1

/* Define to 1 if you have the <string.h> header file. */
#define HAVE_STRING_H 1

/* Define to 1 if you have the <sys/ioctl.h> header file. */
#define HAVE_SYS_IOCTL_H 1

/* Define to 1 if you have the <sys/stat.h> header file. */
#define HAVE_SYS_STAT_H 1

/* Define to 1 if you have the <sys/types.h> header file. */
#define HAVE_SYS_TYPES_H 1

/* Define to 1 if you have the <unistd.h> header file. */
#define HAVE_UNISTD_H 1

/* Name of package */
#define PACKAGE "gtkguitune"

/* Define to the address where bug reports for this package should be sent. */
#define PACKAGE_BUGREPORT ""

/* Define to the full name of this package. */
#define PACKAGE_NAME ""

/* Define to the full name and version of this package. */
#define PACKAGE_STRING ""

/* Define to the one symbol short name of this package. */
#define PACKAGE_TARNAME ""

/* Define to the version of this package. */
#define PACKAGE_VERSION ""

/* Define to 1 if you have the ANSI C header files. */
#define STDC_HEADERS 1

/* Version number of package */
#define VERSION "0.7"

/* Define to empty if `const' does not conform to ANSI C. */
/* #undef const */


guitune.h
// guitune.h
//
//    guitune - program for tuning instruments (actually an oscilloscope)
//    Copyright (C) 1999  Florian Berger
//    Email: florian.berger@jk.uni-linz.ac.at
//
//    This program is free software; you can redistribute it and/or modify
//    it under the terms of the GNU General Public License Version 2 as
//    published by the Free Software Foundation;
//
//    This program is distributed in the hope that it will be useful,
//    but WITHOUT ANY WARRANTY; without even the implied warranty of
//    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//    GNU General Public License for more details.
//
//    You should have received a copy of the GNU General Public License
//    along with this program; if not, write to the Free Software
//    Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
//
//

//#include <gtkmm/window.h>
#include <gtkmm/container.h>
#include <gtkmm/table.h>
#include <gtkmm/label.h>
#include <gtkmm/adjustment.h>
#include <gtkmm/checkbutton.h>

#include "osziview.h"
#include "logview.h"
#include "lcdview.h"


class MainWidget : public Gtk::Table
{
 public:
   MainWidget();
   ~MainWidget();
   int sampnr;
   int sampfreq;
   double sampfreq_exact;
   int blksize;
   int trig1;
   int trig2;
   int    note_0t;
   int    note_ht;
   double freq_0t;
   double freq_ht;
   double lfreq_0t;
   double lfreq_ht;
   void setTuningNorm();
   void setTuningWien();
   void setTuningPhys();
   void setTuningEqui();
   void setTuningNat();
   void setScaleUS();
   void setScaleUSAlt();
   void setScaleGE();
   void setScaleGEAlt();
   void showOszi();
   void hideOszi();
   void showLogView();
   void hideLogView();
   double getTrigger();
// public slots:
     void setSampFreqVal( int sampfreq );
     void setSampFreq();
     void setSampNrVal( int sampnr );
     void setSampNr();
     void setTrigger();
     void setTriggerVal(double value);
     void setAdaptive();
     void setAdaptiveVal( int adaptive );
     void setDSPName(const char* name);
// signals:
//   void signalSampFreqChanged();
//   void signalSampNrChanged();
//   void signalTriggerChanged();

   Gtk::Adjustment  *  i_pAdjSF;  // sample freq
   Gtk::Adjustment  *  i_pAdjSN;  // sample numner
   Gtk::Adjustment  *  i_pAdjTR;  // trigger
   Gtk::CheckButton *  i_pCBAdapt;  // adaptive scaling
   
private:
   OsziView         *  oszi;
   LogView          *  logview;
//   Gtk::Label       *  freqview;
//   Gtk::Label       *  nfreqview;
   LCDView          *  freqview;
   LCDView          *  nfreqview;
   int oszi_height;
   short int sample[64000];
   double freqs[12];
   double lfreqs[12];
   char dsp_devicename[100];
   int  i_sampfmt;
   int  audio_fd;
   sigc::connection audio_connection;

   bool trig;
   int trigpos;
   int got_bytes;

// private slots:
   bool proc_audio(Glib::IOCondition cond);

   void init_audio();
   void uninit_audio();
};


guitune.cc
// guitune.cc
//
//    guitune - program for tuning instruments (actually an oscilloscope)
//    Copyright (C) 1999  Florian Berger
//    Email: florian.berger@jk.uni-linz.ac.at
//
//    This program is free software; you can redistribute it and/or modify
//    it under the terms of the GNU General Public License Version 2 as
//    published by the Free Software Foundation;
//
//    This program is distributed in the hope that it will be useful,
//    but WITHOUT ANY WARRANTY; without even the implied warranty of
//    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//    GNU General Public License for more details.
//
//    You should have received a copy of the GNU General Public License
//    along with this program; if not, write to the Free Software
//    Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
//
//

//#include <iostream.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <sys/ioctl.h>
#include <linux/soundcard.h>

#include <gtkmm/main.h>
#include <gtkmm/table.h>
#include <gtkmm/spinbutton.h>
#include <gtkmm/style.h>
#include <gtkmm/frame.h>

#include <unistd.h>
#include <stdio.h>
#include <time.h>
#include <stdlib.h>
#include <string.h>
#include <errno.h>

#include <math.h>

#include "guitune.h"

#include "resources.h"



//globally
double KAMMERTON, KAMMERTON_LOG;

void MainWidget::setTuningNorm()
{
    KAMMERTON=KAMMERTON_NORM;
    KAMMERTON_LOG=KAMMERTON_LOG_NORM;
    logview->invalidate();
}

void MainWidget::setTuningWien()
{
    KAMMERTON=KAMMERTON_WIEN;
    KAMMERTON_LOG=KAMMERTON_LOG_WIEN;
    logview->invalidate();
}

void MainWidget::setTuningPhys()
{
    KAMMERTON=KAMMERTON_PHYS;
    KAMMERTON_LOG=KAMMERTON_LOG_PHYS;
    logview->invalidate();
}


void MainWidget::setTuningEqui()
{
    logview->set_nat_tuning(false);
}


void MainWidget::setTuningNat()
{
    logview->set_nat_tuning(true);
}


void MainWidget::setScaleUS()
{
    logview->setScale(LogView::us_scale);
}

void MainWidget::setScaleUSAlt()
{
    logview->setScale(LogView::us_scale_alt);
}

void MainWidget::setScaleGE()
{
    logview->setScale(LogView::german_scale);
}

void MainWidget::setScaleGEAlt()
{
    logview->setScale(LogView::german_scale_alt);
}



void MainWidget::showLogView()
{
    logview->show();
}

void MainWidget::hideLogView()
{
    logview->hide();
}

void MainWidget::showOszi()
{
    oszi->show();
}

void MainWidget::hideOszi()
{
    oszi->hide();
}

double MainWidget::getTrigger()
{
    return(oszi->getTrigFact());
}

void MainWidget::setSampFreqVal( int sampfreq )
{
    i_pAdjSF->set_value(sampfreq);
}


void MainWidget::setSampFreq()
{
    sampfreq = (int) i_pAdjSF->get_value();
//    timer->stop();
    init_audio();
//    timer->start(0);
    oszi->setSampleFreq(sampfreq_exact);
//    emit signalSampFreqChanged();
}


void MainWidget::setSampNrVal( int sampnr )
{
    i_pAdjSN->set_value(sampnr);
}


void MainWidget::setSampNr()
{
    sampnr = (int) i_pAdjSN->get_value();
//    emit signalSampNrChanged();
}


void MainWidget::setTrigger()
{
    oszi->setTrigFact(i_pAdjTR->get_value());
//    emit signalTriggerChanged();
}


void MainWidget::setTriggerVal(double value)
{
//    oszi->setTrigFact(value);
    i_pAdjTR->set_value(value);
//    i_pAdjTR->changed();
//    emit signalTriggerChanged();
}


void MainWidget::setAdaptive()
{
    bool active;
    active = i_pCBAdapt->get_active();
    oszi->setAdaptive((int)active);
}


void MainWidget::setAdaptiveVal( int adaptive )
{
    i_pCBAdapt->set_active((bool)adaptive);
}



void MainWidget::setDSPName(const char *name)
{
//    timer->stop();
    strcpy(dsp_devicename,name);
    init_audio();
//    timer->start(0);
}

void MainWidget::init_audio()
{
   printf("initializing audio at %s\n",dsp_devicename);

   uninit_audio();
   
   audio_fd = open(dsp_devicename, O_RDONLY);
   if (audio_fd == -1) {
      perror(dsp_devicename);
      exit(1);
   }
   fcntl(audio_fd,F_SETFD,FD_CLOEXEC);

//   ioctl(audio_fd, SNDCTL_DSP_RESET, 0);
   if( strcmp(dsp_devicename,"/dev/stdin")==0 ){

      printf("reading data from stdin\n");
   		blksize=32;
      printf("  blocksize=%d\n",blksize);
      printf("  sampfreq=%d\n",sampfreq);
      sampfreq_exact=sampfreq;

   }else{

      ioctl(audio_fd, SNDCTL_DSP_SETDUPLEX, 0);

      {int caps;
          ioctl( audio_fd, SNDCTL_DSP_GETCAPS, &caps );
          printf("OSS-Version %d\n", caps & DSP_CAP_REVISION );
       		printf("  DUPLEX   = %X\n",caps & DSP_CAP_DUPLEX   );
          printf("  REALTIME = %X\n",caps & DSP_CAP_REALTIME );
          printf("  BATCH    = %X\n",caps & DSP_CAP_BATCH    );
          printf("  COPROC   = %X\n",caps & DSP_CAP_COPROC   );
          printf("  TRIGGER  = %X\n",caps & DSP_CAP_TRIGGER  );
          printf("  MMAP     = %X\n",caps & DSP_CAP_MMAP     );
   		}

   		blksize=8;  // 2^8 = 256
  		ioctl(audio_fd, SNDCTL_DSP_SETFRAGMENT, &blksize);
//   ioctl(audio_fd, SNDCTL_DSP_SETBLKSIZE, &blksize);

      ioctl(audio_fd, SNDCTL_DSP_GETBLKSIZE, &blksize);
      printf("blocksize=%d\n",blksize);
   //blksize=4096;

      ioctl(audio_fd, SNDCTL_DSP_SYNC, NULL);
//   int a_sampsize = 8;
//      int a_sampsize = AFMT_U8;
      i_sampfmt = AFMT_S16_LE;
      ioctl(audio_fd, SNDCTL_DSP_SAMPLESIZE, &i_sampfmt);
      int a_stereo = 0;
      ioctl(audio_fd, SNDCTL_DSP_STEREO, &a_stereo);

      int a_speed = sampfreq;
      printf("sampfreq=%d\n",sampfreq);
      ioctl(audio_fd, SNDCTL_DSP_SPEED, &a_speed);
      ioctl(audio_fd, SOUND_PCM_READ_RATE, &sampfreq);
      printf("sampfreq=%d\n",sampfreq);
      sampfreq_exact=sampfreq;

   }

   //int mixer = open("/dev/mixer", O_RDONLY, 0);
   //int vol=0xFFFF;
   //ioctl(mixer, SOUND_MIXER_WRITE_MIC, &vol);
   //printf("MIC-Volume=%d\n",vol);
   //close(mixer);

// printf("close(audio_fd)=%d\n",close(audio_fd));
// printf("close(audio_fd)=%d\n",close(audio_fd));

   trig = false;
   got_bytes = 0;

   audio_connection = Glib::signal_io()
      .connect(sigc::mem_fun(this, &MainWidget::proc_audio),
	       audio_fd, Glib::IO_IN | Glib::IO_HUP);
}

void MainWidget::uninit_audio()
{
   audio_connection.disconnect();

   if (audio_fd != -1)
   {
      ::close(audio_fd);
      audio_fd = -1;
   }
}

MainWidget::MainWidget()
    : Gtk::Table(7, 4)
{
    int i;
   //InitAudio();
//   setMinimumSize(490,240);

    set_row_spacings( 12 );
    set_col_spacings( 12 );

    KAMMERTON=KAMMERTON_NORM;
    KAMMERTON_LOG=KAMMERTON_LOG_NORM;

    strcpy(dsp_devicename,"/dev/dsp");

    sampnr=1024;
    sampfreq=11048;
    audio_fd = -1;
    init_audio();
    printf("Audiodriver initialized\n");

    freqs[0]=KAMMERTON; lfreqs[0]=KAMMERTON_LOG;
    for(i=1;i<12;i++){
        freqs [i] = freqs [i-1] * D_NOTE;
        lfreqs[i] = lfreqs[i-1] + D_NOTE_LOG;
    }

    oszi = new OsziView;
    oszi->setSampleFreq(sampfreq_exact);
    logview = new LogView;
    
//    Gtk::Style style = new Gtk::Style( new void );
//    Gdk_Font font("-*-helvetica-bold-r-*-*-15-*-*-*-*-*-*-*");
    //    get_style()->set_font( font );
    
//    freqview  = new Gtk::Label("freq");
//    nfreqview = new Gtk::Label("nfreq");
    freqview  = new LCDView("");
    freqview->set_size_request(100,20);
    nfreqview = new LCDView("");
    nfreqview->set_size_request(100,20);
    
//    nfreqview->set_style( style );
    
    
    i_pAdjSF = new Gtk::Adjustment(
                               11024.0,  // value
                                5000.0,  // lower
                               48000.0,  // upper
                                   1.0,  // step inc
                                 100.0,  // page inc
                                     0   // page size
                                  );
    
    i_pAdjSN = new Gtk::Adjustment(
                                1024.0,  // value
                                 100.0,  // lower
                               10000.0,  // upper
                                   1.0,  // step inc
                                 100.0,  // page inc
                                     0   // page size
                                  );
    
    i_pAdjTR = new Gtk::Adjustment(
                                   0.6,  // value
                                   0.0,  // lower
                                   1.0,  // upper
                                  0.05,  // step inc
                                   0.1,  // page inc
                                     0   // page size
                                  );

    attach( *oszi, 0, 1, 0, 5 );

    attach( *manage( new Gtk::Label("Freq:") ),
	    1, 2, 1, 2,
	    Gtk::AttachOptions(), Gtk::AttachOptions() );
    attach( *freqview, 2, 3, 1, 2,
	    Gtk::FILL, Gtk::AttachOptions() );

    attach( *manage( new Gtk::Label("Tune:") ),
	    1, 2, 2, 3,
	    Gtk::AttachOptions(), Gtk::AttachOptions() );
    attach( *nfreqview, 2, 3, 2, 3,
	    Gtk::FILL, Gtk::AttachOptions() );

    Gtk::Frame *adj_frame = new Gtk::Frame("Adjust:");
    
    Gtk::Table *adj_table = new Gtk::Table(6, 4);

    adj_table->set_row_spacings( 12 );
    adj_table->set_col_spacings( 12 );

    adj_frame->add( *manage( adj_table ) );

    adj_table->attach( *manage( new Gtk::Label("SF:") ),
		       1, 2, 1, 2,
		       Gtk::FILL, Gtk::AttachOptions() );

    Gtk::SpinButton *spin_sf = new Gtk::SpinButton( *i_pAdjSF );

    adj_table->attach( *manage( spin_sf ),
		       2, 3, 1, 2 );

    adj_table->attach( *manage( new Gtk::Label("SN:") ),
		       1, 2, 2, 3,
		       Gtk::AttachOptions(), Gtk::AttachOptions() );

    Gtk::SpinButton *spin_sn = new Gtk::SpinButton( *i_pAdjSN );

    adj_table->attach( *manage( spin_sn ),
		       2, 3, 2, 3 );

    adj_table->attach( *manage( new Gtk::Label("Trig:") ),
		       1, 2, 3, 4,
		       Gtk::AttachOptions(), Gtk::AttachOptions() );

    Gtk::SpinButton *spin_tr = new Gtk::SpinButton( *i_pAdjTR );

    adj_table->attach( *manage( spin_tr ),
		       2, 3, 3, 4 );

    i_pCBAdapt = new Gtk::CheckButton( "Adapt" );

    adj_table->attach( *manage( i_pCBAdapt ),
		       2, 3, 4, 5 );

    attach( *manage( adj_frame ),
	    1, 3, 3, 4,
	    Gtk::FILL, Gtk::FILL );

    attach( *manage( new Gtk::Label( "" ) ),
	    1, 3, 4, 5, Gtk::AttachOptions() );

    attach( *logview,
	    0, 3, 5, 6,
	    Gtk::EXPAND | Gtk::FILL, Gtk::AttachOptions() );

    logview->set_size_request(0,86);
       
    spin_tr->set_digits(2);
       
    i_pAdjSF->signal_value_changed().connect( sigc::mem_fun( this, &MainWidget::setSampFreq ) );
    i_pAdjSN->signal_value_changed().connect( sigc::mem_fun( this, &MainWidget::setSampNr ) );
    i_pAdjTR->signal_value_changed().connect( sigc::mem_fun( this, &MainWidget::setTrigger ) );

    i_pAdjSF->signal_value_changed().connect( sigc::mem_fun( oszi, &OsziView::invalidate ) );
    i_pAdjSN->signal_value_changed().connect( sigc::mem_fun( oszi, &OsziView::invalidate ) );

    i_pCBAdapt->signal_toggled().connect( sigc::mem_fun( this, &MainWidget::setAdaptive ) );
}


bool MainWidget::proc_audio(Glib::IOCondition cond)
{int i,j,r,got_samples;
 unsigned char *c;
 short int *s;
 unsigned char * sample_uc;
 short int *     sample_s16le;
 double ldf,mldf;
 char str[50];

 sample_uc    = (unsigned char *)sample;
 sample_s16le = (short int *)sample;
 c = sample_uc;
 s = sample_s16le;

 r = read(audio_fd, c + got_bytes, blksize);

 if (r == -1)
 {
    g_critical("error reading audio: %s", strerror(errno));
    uninit_audio();
    return false;
 }
 if (r == 0)
 {
    g_critical("EOF on audio");
    uninit_audio();
    return false;
 }

 got_bytes += r;

 got_samples = got_bytes;
 if (i_sampfmt == AFMT_S16_LE)
    got_samples /= 2;

 if (!trig)
 {
   // Look for a trigger
   trigpos=0;

   if(i_sampfmt==AFMT_U8){
       for( i=0; i<got_samples && Abs(c[i]-128)<2; i++ );
   } else if(i_sampfmt==AFMT_S16_LE) {
       for( i=0; i<got_samples && Abs(s[i])<256; i++ );
   }
   //i=-1;
   j=0; trig=0;
   if (i<got_samples) {
      if(i_sampfmt==AFMT_U8){
          for( ; i<got_samples-1; i++ )   /* n-1 because of POSTRIG uses i+1 */
              if ( POSTRIG(c,i) ) { trig = true; trigpos=i; }
      }else if(i_sampfmt==AFMT_S16_LE){
          for( ; i<got_samples-1; i++ )   /* n-1 because of POSTRIG uses i+1 */
              if ( POSTRIG_S16(s,i) ) { trig = true; trigpos=i; }
      }
   }
   // If we didn't find a trigger then discard the bytes we read
   if (!trig)
      got_bytes = 0;
 }
 else if (got_samples - trigpos >= sampnr)
 {
      if(i_sampfmt==AFMT_U8)
	 oszi->setSampleData( sample_uc + trigpos, sampnr );
      else if(i_sampfmt==AFMT_S16_LE)
	 oszi->setSampleData( sample_s16le + trigpos, sampnr );
      
      freq_0t = (double)sampfreq*oszi->get_freq();
      lfreq_0t = log(freq_0t);
      while ( lfreq_0t < lfreqs[0]-D_NOTE_LOG/2.0 ) lfreq_0t+=LOG_2;
      while ( lfreq_0t >= lfreqs[0]+LOG_2-D_NOTE_LOG/2.0 ) lfreq_0t-=LOG_2;
      mldf=D_NOTE_LOG; note_0t=0;
      for( i=0; i<12; i++ ){
          ldf = fabs(lfreq_0t-lfreqs[i]);
          if (ldf<mldf) { mldf=ldf; note_0t=i; }
      }
      logview->change_lfreq(lfreq_0t);
      sprintf(str,"%0.3f",freq_0t);
      freqview->set(str);
      double nfreq_0t=freqs[note_0t];
      while( nfreq_0t/freq_0t > D_NOTE_SQRT ) nfreq_0t/=2.0;
      while( freq_0t/nfreq_0t > D_NOTE_SQRT ) nfreq_0t*=2.0;
      sprintf(str,"%0.3f",nfreq_0t);
      nfreqview->set(str);
      
//      printf("Note: %s (%lfHz) Freq=%lf\n",
//	        note[note_0t],freqs[note_0t],freq_0t);

      trig = false;
      got_bytes = 0;
 }
   
   return true;
}

MainWidget::~MainWidget()
{
   uninit_audio();
}
